src/openrct2/core/FileIndex.hpp

```diff
@@ -178,32 +180,94 @@ class FileIndex
         return ScanResult(stats, files);
     }
 
-    std::vector<TItem> Build(const ScanResult &scanResult) const
+    void BuildRange(const ScanResult &scanResult,
+                    size_t rangeStart,
+                    size_t rangeEnd,
+                    std::vector<TItem>& items,
+                    std::atomic<size_t>& processed,
+                    std::mutex& printLock) const
     {
-        std::vector<TItem> items;
-        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
-
-        auto startTime = std::chrono::high_resolution_clock::now();
-        // Start at 1, so that we can reach 100% completion status
-        size_t i = 1;
-        for (auto filePath : scanResult.Files)
+        items.reserve(rangeEnd - rangeStart);
+        for (size_t i = rangeStart; i < rangeEnd; i++)
         {
-            Console::WriteFormat("File %5d of %d, done %3d%%\r", i, scanResult.Files.size(), i * 100 / scanResult.Files.size());
-            i++;
-            log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            const auto& filePath = scanResult.Files.at(i);
+
+            if (_log_levels[DIAGNOSTIC_LEVEL_VERBOSE])
+            {
+                std::lock_guard<std::mutex> lock(printLock);
+                log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            }
+
             auto item = Create(filePath);
             if (std::get<0>(item))
             {
                 items.push_back(std::get<1>(item));
             }
+
+            processed++;
         }
+    }
+
+    std::vector<TItem> Build(const ScanResult &scanResult) const
+    {
+        std::vector<TItem> allItems;
+        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+
+        const size_t totalCount = scanResult.Files.size();
+        if (totalCount > 0)
+        {
+            JobPool jobPool;
+            std::mutex printLock; // For verbose prints.
+
+            std::list<std::vector<TItem>> containers;
 
-        WriteIndexFile(scanResult.Stats, items);
+            size_t stepSize = 100; // Handpicked, seems to work well with 4/8 cores.
+
+            std::atomic<size_t> processed = ATOMIC_VAR_INIT(0);
+
+            auto reportProgress = [&]()
+            {
+                const size_t completed = processed;
+                Console::WriteFormat("File %5d of %d, done %3d%%\r", completed, totalCount, completed * 100 / totalCount);
+            };
+
+            for (size_t rangeStart = 0; rangeStart < totalCount; rangeStart += stepSize)
+            {
+                if (rangeStart + stepSize > totalCount)
+                    stepSize = totalCount - rangeStart;
+
+                containers.emplace_back();
```

Comment:
Please leave a note here: `// TODO: change to auto& items = containers.emplace_back() in C++17`